File: Difficulty: Medium/Frog Jump/frog-jump.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    int f(int index, vector<int> &heights,vector<int> &memo){
        if(index==0)return 0;
        if(memo[index]!=-1)return memo[index];
        int left = f(index-1,heights,memo) + abs(heights[index]-heights[index-1]);
        int right=1e9;
        if(index>1)right = f(index-2,heights,memo)+abs(heights[index]-heights[index-2]);
        return memo[index] = min(left,right);
    }
    int minCost(vector<int>& height) {
        // Code here
        int n=height.size();
        vector<int> memo(n,-1);
        return f(n-1, height,memo);
    }
};


//{ Driver Code Starts.

int main() {
    string str;
    getline(cin, str);
    int t = stoi(str);
    while (t--) {
        getline(cin, str);
        stringstream ss(str);
        vector<int> arr;
        int num;
        while (ss >> num) {
            arr.push_back(num);
        }
        Solution ob;
        cout << ob.minCost(arr) << endl;
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Frog Jump/frog-jump.py
#{ 
 # Driver Code Starts
#Initial Template for Python 3
import sys


# } Driver Code Ends

#User function Template for python3
class Solution:
    def minCost(self, height):
        # code here\
        n = len(height)
        memo = [-1] * n

        def fn(ind):
            if ind==0:
                return 0
            if memo[ind]!=-1:
                return memo[ind]
            oneStep = abs(height[ind]-height[ind-1]) + fn(ind-1)
            twoStep = 1e8
            if ind>=2:
                twoStep =  abs(height[ind]-height[ind-2]) + fn(ind-2)
            memo[ind]= min(oneStep,twoStep)
            return memo[ind]
            
            
        return fn(n-1)


#{ 
 # Driver Code Starts.

if __name__ == "__main__":
    input_lines = sys.stdin.read().splitlines()
    
    pointer = 0
    
    t = int(input_lines[pointer].strip())
    pointer += 1
    
    for _ in range(t):
        arr = list(map(int, input_lines[pointer].strip().split()))
        pointer += 1
        
        ob = Solution()
        print(ob.minCost(arr))
        print("~")

# } Driver Code Ends